File: app/src/main/java/com/maciega/bartosz/crawl/model/Spider.java
package com.maciega.bartosz.crawl.model;

import android.content.Context;
import android.os.AsyncTask;
import android.util.Log;
import android.widget.Toast;

import com.maciega.bartosz.crawl.storage.DbStorage;
import com.maciega.bartosz.crawl.storage.DbTransactionListener;
import com.maciega.bartosz.crawl.storage.Storage;
import com.maciega.bartosz.crawl.storage.StorageTransactionResult;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

import rx.Observable;

/**
 * Created by bartoszmaciega on 17/10/16.
 */

public class Spider implements DbTransactionListener {
    private int max_pages = 20;
    private Set<String> visitedPages = new HashSet<>();
    private List<String> pagesToVisit = new LinkedList<>();
    private Context context;
    private CrawlListener crawlListener;


    public Spider(Context context, int pagesToSearch, CrawlListener crawlListener) {
        this.max_pages = pagesToSearch > 0 ? pagesToSearch : 20;
        this.context = context;
        this.crawlListener = crawlListener;
    }


    private synchronized String nextUrl() {
        String nextUrl;

        synchronized (pagesToVisit) {
            do {
                nextUrl = pagesToVisit.remove(0);
            } while (this.visitedPages.contains(nextUrl));
        }

        this.visitedPages.add(nextUrl);
        return nextUrl;
    }


    public void search(String url) {

        SpiderAsync spiderAsync = new SpiderAsync();
        spiderAsync.execute(url);
    }

    @Override
    public void onResult(StorageTransactionResult result) {

    }


    class SpiderAsync extends AsyncTask<String, String, Void> implements CrawlListener, DbTransactionListener {


        @Override
        protected Void doInBackground(String... params) {
            String url = params[0];
            pagesToVisit.add(url);

            while (visitedPages.size() < max_pages && !pagesToVisit.isEmpty()) {
                try {
                    String currentUrl;
                    CrawlerItself crawler = new CrawlerItself(this);
                    currentUrl = nextUrl();
                    crawler.crawl(currentUrl);
                    publishProgress(currentUrl);

                    pagesToVisit.addAll(crawler.getLinks());
                } catch (Throwable e) {
                    return null;
                }

            }


            return null;
        }

        @Override
        protected void onProgressUpdate(String... values) {
            super.onProgressUpdate(values);
            crawlListener.onCrawled(values[0]);
        }


        @Override
        protected void onPostExecute(Void aVoid) {
            super.onPostExecute(aVoid);
            Storage storage = new DbStorage(context);
            storage.saveList(UrlsMapper.INSTANCE.convert(visitedPages));
            Toast.makeText(context, "searching finished", Toast.LENGTH_SHORT).show();
        }

        @Override
        public void onCrawled(String pageUrl) {

        }

        @Override
        public void onResult(StorageTransactionResult result) {
            if (result.isSuccess()) {
                Log.d("success", "success");
            }
        }
    }


}

File: app/src/main/java/com/maciega/bartosz/crawl/model/Spider.kt
package com.maciega.bartosz.crawl.model

import android.content.Context
import android.os.AsyncTask
import android.util.Log
import android.widget.Toast

import com.maciega.bartosz.crawl.storage.DbStorage
import com.maciega.bartosz.crawl.storage.DbTransactionListener
import com.maciega.bartosz.crawl.storage.Storage
import com.maciega.bartosz.crawl.storage.StorageTransactionResult
import io.realm.RealmObject

import java.util.ArrayList
import java.util.Arrays
import java.util.HashSet
import java.util.LinkedList

import rx.Observable

/**
 * Created by bartoszmaciega on 17/10/16.
 */

class Spider(private val context: Context, pagesToSearch: Int, private val crawlListener: CrawlListener) : DbTransactionListener {
    private var max_pages = 20
    private val visitedPages = HashSet<String>()
    private val pagesToVisit = LinkedList<String>()


    init {
        this.max_pages = if (pagesToSearch > 0) pagesToSearch else 20
    }


    @Synchronized private fun nextUrl(): String {
        var nextUrl: String = ""

        synchronized(pagesToVisit) {
            do {
                nextUrl = pagesToVisit.removeAt(0)
            } while (this.visitedPages.contains(nextUrl))
        }

        this.visitedPages.add(nextUrl)
        return nextUrl
    }


    fun search(url: String) {

        val spiderAsync = SpiderAsync()
        spiderAsync.execute(url)
    }

    override fun onResult(result: StorageTransactionResult<*>) {

    }


    internal inner class SpiderAsync : AsyncTask<String, String, Void>(), CrawlListener, DbTransactionListener {


        override fun doInBackground(vararg params: String): Void? {
            val url = params[0]
            pagesToVisit.add(url)

            while (visitedPages.size < max_pages && !pagesToVisit.isEmpty()) {
                try {
                    val currentUrl: String
                    val crawler = CrawlerItself(this)
                    currentUrl = nextUrl()
                    crawler.crawl(currentUrl)
                    publishProgress(currentUrl)

                    pagesToVisit.addAll(crawler.getLinks())
                } catch (e: Throwable) {
                    return null
                }

            }


            return null
        }

        override fun onProgressUpdate(vararg values: String) {
            super.onProgressUpdate(*values)
            crawlListener.onCrawled(values[0])
        }


        override fun onPostExecute(aVoid: Void) {
            super.onPostExecute(aVoid)
            val storage = DbStorage(context)
            storage.saveList<RealmObject>(UrlsMapper.convert(visitedPages))
            Toast.makeText(context, "searching finished", Toast.LENGTH_SHORT).show()
        }

        override fun onCrawled(pageUrl: String) {

        }

        override fun onResult(result: StorageTransactionResult<*>) {
            if (result.isSuccess) {
                Log.d("success", "success")
            }
        }
    }


}
